### PR TITLE
Configure dev environment with `flake.nix` and `pyproject.toml`

### PR DESCRIPTION
**Pull request by jackdbd**
This PR introduces a `flake.nix` to automatically configure a development environment.

I also updated the `pyproject.toml` to use [poetry](https://python-poetry.org/docs/) instead of setuptools as the build system, and organized python dependencies in [dependency groups](https://python-poetry.org/docs/managing-dependencies/).

### CODE REVIEW

File: benchmarks/symbol_size_benchmark.py [hunk 1]
```diff
@@ -61,40 +61,47 @@ def sqlelf_memoized_benchmark(sqlite_database: str, num_functions: int) -> None:
...
+                    ).repeat(repeat=10, number=1)
+                )
             )
-        )
 
-pprint.pprint(data)  # noqa: T203
+    pprint.pprint(data)  # noqa: T203
+
+
+if __name__ == "__main__":
```
Comment (fzakaria) [inline benchmarks/symbol_size_benchmark.py:106]: this just better practice or needed for the flake?

**Comment by fzakaria** [comment]
This looks like it's in the right direction and I'm in fact on NixOS now also, so I need to figure out something but the challenge I have with this:
1 - it makes publishing to PyPI difficult
2 - it changes a lot with the move to Poetry

What about just a shell.nix file?

File: benchmarks/symbol_size_benchmark.py (hunk 1 above)
Comment (jackdbd) [reply benchmarks/symbol_size_benchmark.py:106]: It's a very common python pattern. The conditional block prevents the code from running when the python module is imported by other modules. The only time the code runs is when it is executed as a script.

I think I had to add it when calling the script with `poetry run`. I'm not 100% sure though, I wrote it a couple of weeks ago.

**Comment by jackdbd** [comment]
Ah right, I haven't thought about publishing on PyPI. I just couldn't figure out how to setup a development environment using the provided project.toml. That's why I updated that file and migrated the build system from setuptools to poetry.

Do you think it's better to keep the CI setup (Ubuntu) separated from the dev setup (shell.nix, flake.nix, devenv.nix, etc)?

**Comment by fzakaria** [comment]
I pushed the nix flake.
We can close this now -- thank you for the inspiration.

**Pull request closed without merging**